### DISCUSSION TITLE
Manual vectorization of the spreader

### DISCUSSION THREAD

**Discussion by DiamonDinoia** [flatironinstitute/finufft#452] [Ideas]
Manually vectorizing Horner and the spreader can give speedup sometimes doubling the performance. In particular clang vectorizer seems not to be aggressive leaving performance on the table. 
The vectorization is based on the spreader width (`ns`). `ns` is a runtime variable, hence it is not possible to rely on it to do any fine vectorization. However, the supported range for `ns` is `[2, 16]`. Under these assumptions it is possible to generate compile time kernels (using c++17 and templates) for each value these values. Then, at runtime the correct kernel can be selected. This improves the performance by a large margin (up to 100% in certain cases, more details below). The drawback is more verbose and less clear code. By manual vectorizing it is possible to do extra optimizations that rely on knowing that some values are `don't cares` and even if garbage or zeros are written in it, it does not impact the final result as these are ignored. Another drawback is that padding is required in certain places to guarantee that no out-of-bounds memory accesses take place.
The vectorization relies on xsimd [xsimd](https://github.com/xtensor-stack/xsimd). The reason is that intrinsics are not portable, this library is well maintained and supports arm, mac, windows and linux.

The source code is available here: [fork](https://github.com/DiamonDinoia/finufft/tree/spreader-vectorization-v2)
I would greatly appreciate any suggestions to improve it.

EDIT: 06 Jun 2024: I updated the branch to the most recent one.

Performance measurements:

Older AMD processor (does not have AVX512)
![AMDEPYC774264-CoreProcessor-1D](https://github.com/flatironinstitute/finufft/assets/9217729/d954510c-aecd-44e7-93ca-8c4ad510a43d)
![AMDEPYC774264-CoreProcessor-2D](https://github.com/flatironinstitute/finufft/assets/9217729/461fc365-e88a-4d4b-8694-a0000b72e863)
![AMDEPYC774264-CoreProcessor-3D](https://github.com/flatironinstitute/finufft/assets/9217729/9f3698e3-b8ac-418d-9063-631158bc2e02)
Performance improved in all cases especially in 1D and 2D where it doubles. 3D minor improvements.

Older Intel Gold (supports AVX512)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/6a0596b6-ce4a-4e1e-8a10-96a54b001ac8)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/d4ec0fe6-57b7-4e5b-9cdb-08902fa4b103)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/b541667e-48a3-49f5-8f5f-30b09ae512c0)
Similar to the AMD machine but higher improvements on 3D

Recent Intel xeon CPU (supports avx512)
![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/17dd6440-9ed4-4b37-87a4-4bb346f48a98)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/99c816d2-71b8-4c33-80c0-40a4a82b7ed7)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/4718a8da-58e3-45d2-9052-2b1f9670d2ca)
Improvements across the board but some minor slowdown in 3D with `ns = [13-15]`.

It is worth noting that manually vectorized code with clang usually offers best performance, auto vectorization in clang is much slower than gcc.

**Comment by DiamonDinoia** [comment]
I updated the branch [fork](https://github.com/DiamonDinoia/finufft/tree/spreader-vectorization-v2) with comments and some minor optimizations. @lu1and10 provided some optimized Horner code that is even faster than my version.

The results are:
On older CPUs:
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/e2870bed-ab4f-4015-9abf-a274b68e3c9c)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/3c2cd2cb-1460-4c2b-83d6-a8d9c8c70e7d)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/e5019a50-d66a-4f60-bd15-a7529bcd3d0a)
and on newer:
![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/b19dab4c-39df-4cab-b36a-8461ac666de6)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/849c92d1-3bc6-4fe1-96d9-665d4f81c92e)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/627256be-769c-4b12-a346-f8f37f9e5b6c)

**Comment by DiamonDinoia** [comment]
These are the results for interpolation, moved ns (aka w) at compile time, generated kernels and dispatched them. It also uses the explicitly vectorized horner:
Older CPU:
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/8fa6cbf4-814b-41ed-bb7a-26ec3ad8d803)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/127a2da6-453f-48dd-9443-ef9cdfe36eb6)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/22ca6bc0-ce5d-42bd-8ce3-96fda502e72a)
Newer CPU:
![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/321ac75e-cfa5-4110-8d87-6b4ee19ce1c2)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/48e77466-8e31-4877-8df1-335c4dea7e36)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/b0f460cb-0615-456f-a950-fcfdda2d1611)


Overall is faster. Although, explicit vectorization should help in this case too.

**Comment by ahbarnett** [comment]
Both spread and interp comparison (master vs spreader-vectorization-v2 commit 3e243269), on Gcc11.4, AMD laptop 5700U ryzen2, one thread. 3D for interp is a great speed-up (1D less dramatic):

![display](https://github.com/flatironinstitute/finufft/assets/10197194/8d8d5ffc-7371-4429-9004-18c029f7a714)
![display](https://github.com/flatironinstitute/finufft/assets/10197194/317890f9-374b-4da2-8b58-93cda61f9687)

**Reply by ahbarnett** [reply]
This is done via `perftest/compare_spreads.jl`

**Comment by DiamonDinoia** [comment]
I manually vectorized the interpolation. Results are below. 
I noticed large improvement in 1D. Minor improvements in 1D sometimes gets a a bit worse (eg. with ns=2). In 3D it gets slower for ns<=4, then it gets faster until ns = 13. From 13 to 16 is slightly slower
Old xeon gold:
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/72e89524-4eed-4606-b559-32e166c18167)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/247788fc-b5fe-4485-b2f5-de18f4ca6c19)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/e8025315-138a-49c7-8173-3aa43e5a5660)
New Xeon w5-3435x
![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/655eb16b-3879-4dca-a467-7e038414e6f5)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/5d30c3e0-c17f-478f-855e-6f6d26e8b517)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/1969d75d-e661-4c82-9d8e-3bb3fce327db)

**Comment by DiamonDinoia** [comment]
Tuning the compiler flags leads to some interesting results:

Xeon Gold:

![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/be733755-80c4-47ad-966e-1e2cf43faa54)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/92977aa5-49d8-4f5e-815d-f51127b96aaf)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/99034037-72f4-4ed5-9ff6-f19388b61f50)

Clang gets better in 1D but gcc gets way worse. In 2D/3D clang gets way worse but the other compilers gets faster.


Intel W5: 
![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/d4a804b8-6c12-485a-938b-8612a1378789)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/227fb688-724e-4eb2-87cb-b23a07e36a9b)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/e1cea712-9bd2-4f94-bc4a-9acc118c71be)



There is not much difference in 1D compared to previous results. In 2D and 3D all the compilers improve and gets to the level of gcc 13 which was the fastest before.

EDIT:

![AMDEPYC774264-CoreProcessor-1D](https://github.com/flatironinstitute/finufft/assets/9217729/b1aae568-f80b-4e17-ac74-54f79cc32c55)
![AMDEPYC774264-CoreProcessor-2D](https://github.com/flatironinstitute/finufft/assets/9217729/e7ce0679-6b22-4010-bf7a-3584b85e72e2)
![AMDEPYC774264-CoreProcessor-3D](https://github.com/flatironinstitute/finufft/assets/9217729/f2841aed-d5d9-45f2-9373-b9a587e06658)

Added test on the old AMD machine. IN this case the performance is better on average. Clang is slower than gcc in most cases.

**Comment by ahbarnett** [comment]
For my AMD 5700U ryzen laptop, 1-thread, GCC 11.4. 1D is much faster, but low-acc double-prec 2d and 3d are slower (although single-prec 2d and 3d is faster).  Blue is master of 7-1-24;  yellow is Marco's fork, interp-vectorization branch:

![7-1-24-vs-ivec_gcc114_5700U_nthr1_interp_M1 0e7_N1 0e6](https://github.com/flatironinstitute/finufft/assets/10197194/b85d17b7-3d15-43ab-bb67-5ee2166de306)

**Reply by ahbarnett** [reply]
flags are as in makefile. Done with perftest/spreadinterpnd{f} as driven by compare_spreads.jl

It could be argued that low-prec double slowdown doesn't matter since the user should switch to single-prec there :)

**Comment by DiamonDinoia** [comment]
I fixed some things and re-run the tests.

TLDR: On all CPUs except one manual vectorization improves performance.

Old AMD ROME:

![AMDEPYC774264-CoreProcessor-1D](https://github.com/flatironinstitute/finufft/assets/9217729/2483a6f4-4bf3-4b85-92f6-ed3c0d400ad7)
![AMDEPYC774264-CoreProcessor-2D](https://github.com/flatironinstitute/finufft/assets/9217729/d94a72d1-87c6-4871-a495-d1680cc37e74)
![AMDEPYC774264-CoreProcessor-3D](https://github.com/flatironinstitute/finufft/assets/9217729/3fff8149-d5f6-41dc-8593-e2c20c2bf754)

Gains across all over the board. 2D and 3D with ns=2 gets a bit slower.

New AMD Genoa:

![AMDEPYC9474F48-CoreProcessor-1D](https://github.com/flatironinstitute/finufft/assets/9217729/7d14bab8-8c2e-4373-a71f-b346e7cda930)
![AMDEPYC9474F48-CoreProcessor-2D](https://github.com/flatironinstitute/finufft/assets/9217729/40c4a753-5197-4e20-8bac-4b2a733b9e56)
![AMDEPYC9474F48-CoreProcessor-3D](https://github.com/flatironinstitute/finufft/assets/9217729/7bf5bbec-d6fe-413c-a6bb-742b1041ee30)

Same as the old AMD, in general performance is much better.

Old Intel Xeon Gold:

![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/flatironinstitute/finufft/assets/9217729/80b543b6-c7bc-45c3-a6ab-4e13653f6838)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/flatironinstitute/finufft/assets/9217729/3af637f5-f4ec-4671-9bd9-767b349000d2)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/flatironinstitute/finufft/assets/9217729/811238ce-b8be-4b25-b20b-b32adbab4980)

Weird behavior, I wonder if there is something wrong with the machine. In 1D clang is fast, gcc gets slower. In 2/3D clang dies and gcc gets faster. 

New Intel xeon w5:

![Intel(R)Xeon(R)w5-3435X-1D](https://github.com/flatironinstitute/finufft/assets/9217729/9cc84be2-7666-4e7e-a9a7-4dfc7d072a1c)
![Intel(R)Xeon(R)w5-3435X-2D](https://github.com/flatironinstitute/finufft/assets/9217729/a7c2ad45-e97d-4db1-ad6a-92c897e10cf3)
![Intel(R)Xeon(R)w5-3435X-3D](https://github.com/flatironinstitute/finufft/assets/9217729/f0528b57-3143-402e-a11a-8ddf01a9f722)

Good performance increase across all the board.

**Comment by ahbarnett** [comment]
I ran my benchmarker on my AMD laptop, with the new flags in makefile (interp-vec branch; yellow); against master (blue):

We see spreader slight worsening:

![display](https://github.com/flatironinstitute/finufft/assets/10197194/e7cf0807-04f4-4d26-a4a7-a28e898a4c68)

varying performance for interp (worse for small w, better for larger):

![display](https://github.com/flatironinstitute/finufft/assets/10197194/c97e8c22-f56e-4fe3-9d44-1e7083b4082f)

**Reply by lu1and10** [reply]
I guess Marco's result is using master branch with old compile flags. Are you using master branch also with the new added compile flags? I guess the new optimization flags will also increase the speed of master branch.

**Comment by ahbarnett** [comment]
my master flags are as in master:  `-O3 -funroll-loops -march=native -fcx-limited-range -ffp-contract=fast`
The new flags seem to hurt spreader for w in range 4-7 (code is the same, I think).

**Reply by lu1and10** [reply]
I see, I tried the new flags with master branch, it seems master branch has a good speed up with the new flags for w>7 of interp on FI cluster nodes. I did not check spreader, not sure why the new flags may hurt spreader.

**Comment by DiamonDinoia** [comment]
I attach the results for Gencoeff with 1.25:
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-3D](https://github.com/user-attachments/assets/43f37ef3-f983-4a15-8cf3-128eba71a4bc)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-2D](https://github.com/user-attachments/assets/6fd24db0-fd97-448f-877a-d5ab53646912)
![Intel(R)Xeon(R)Gold6244CPU@3 60GHz-1D](https://github.com/user-attachments/assets/30f2a3e2-7dec-48b2-9ceb-3b091406d15c)

Great improvement in 1D with new coeffs and manual vectorization combined.

**Reply by DiamonDinoia** [reply]
Note: with 1.25 we need to ask for half digit accuracy to use ns = 3, 6, 11, 13, which the test does not do atm.
